src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+  {
+    auto result = env.do_assume(operand, ns);
+    if(result.is_nil())
+      return result;
+    if(result.is_true())
+      or_result = true_exprt();
+  }
+  return or_result;
+}
+
+exprt assume_eq(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto const &equal_expr = to_binary_expr(expr);
+
+  auto left = env.eval(equal_expr.lhs(), ns);
+  auto right = env.eval(equal_expr.rhs(), ns);
+
+  if(left->is_top() || right->is_top())
+    return nil_exprt();
+  if(!is_value(left) || !is_value(right))
+    return nil_exprt();
+
+  auto meet = left->meet(right);
+
+  if(meet->is_bottom())
+    return false_exprt();
+
+  if(is_lvalue(equal_expr.lhs()))
+    env.assign(equal_expr.lhs(), meet, ns);
+  if(is_lvalue(equal_expr.rhs()))
+    env.assign(equal_expr.rhs(), meet, ns);
+  return true_exprt();
+}
+
+exprt assume_noteq(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto const &notequal_expr = to_binary_expr(expr);
+
+  auto left = env.eval(notequal_expr.lhs(), ns);
+  auto right = env.eval(notequal_expr.rhs(), ns);
+
+  if(left->is_top() || right->is_top())
+    return nil_exprt();
+  if(!is_value(left) || !is_value(right))
+    return nil_exprt();
+
+  auto meet = left->meet(right);
+
+  if(meet->is_bottom())
+    return true_exprt();
+
+  return false_exprt();
+}
+
+struct left_and_right_valuest
+{
+  abstract_value_pointert left;
+  abstract_value_pointert right;
+
+  constant_interval_exprt left_interval() const
+  {
+    return left->to_interval();
+  }
+  constant_interval_exprt right_interval() const
+  {
+    return right->to_interval();
+  }
+
+  bool are_good() const
+  {
+    return left != nullptr && right != nullptr;
+  }
+};
+
+left_and_right_valuest eval_operands_as_values(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto const &relationship_expr = to_binary_expr(expr);
+
+  auto left = env.eval(relationship_expr.lhs(), ns);
+  auto right = env.eval(relationship_expr.rhs(), ns);
+
+  if(left->is_top() || right->is_top())
+    return {};
+
+  return {as_value(left), as_value(right)};
+}
+
+exprt assume_less_than(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto operands = eval_operands_as_values(env, expr, ns);
+  if(!operands.are_good())
+    return nil_exprt();
+
+  auto left_lower = operands.left_interval().get_lower();
+  auto right_upper = operands.right_interval().get_upper();
+
+  auto reduced_le_expr =
+    binary_relation_exprt(left_lower, expr.id(), right_upper);
+  return env.eval(reduced_le_expr, ns)->to_constant();
+}
+
+exprt assume_greater_than(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto operands = eval_operands_as_values(env, expr, ns);
+  if(!operands.are_good())
+    return nil_exprt();
+
+  auto left_upper = operands.left_interval().get_upper();
+  auto right_lower = operands.right_interval().get_lower();
+
```

Comment:
Am I right in thinking that this will not resolve the following:

```
int x = nondet();

if (x < 10) {
  assert(x < 20);
}
```

because it does not prune the domain; it just identifies when a branch is infeasible (returning true / false)?

Reply:
That's right, the assert will be UNKNOWN. We only prune the domain in the case of ==

Reply:
It would be really good if we could prune the domain on this case, it really makes a difference for precision and termination.  If there is no overlap then we return true or false (depending on if it is always greater than or never greater than).  If the two overlap (i.e. their meet is non-empty) then you can set the new bounds from the meet.

For example, if `x > y` then you can assign `x` to `[x.upper, y.upper]` and `y` to `[x.lower, y.lower]` (with the off-by-one for `>` vs. `>=`).  It might require me a little thought about how to do this in an abstraction agnostic way.